tests/test_butlerFits.py

```diff
@@ -0,0 +1,214 @@
+        self.assertEqual(inputRecord.getPsfFluxFlag(), outputRecord.getPsfFluxFlag())
+        self.assertEqual(inputTable.getSchema().getAliasMap().get("slot_Centroid"),
+                         outputTable.getSchema().getAliasMap().get("slot_Centroid"))
+        self.assertEqual(inputRecord.getCentroid(), outputRecord.getCentroid())
+        self.assertFloatsAlmostEqual(
+            inputRecord.getCentroidErr()[0, 0],
+            outputRecord.getCentroidErr()[0, 0], rtol=1e-6)
+        self.assertFloatsAlmostEqual(
+            inputRecord.getCentroidErr()[1, 1],
+            outputRecord.getCentroidErr()[1, 1], rtol=1e-6)
+        self.assertEqual(inputTable.getSchema().getAliasMap().get("slot_Shape"),
+                         outputTable.getSchema().getAliasMap().get("slot_Shape"))
+        self.assertFloatsAlmostEqual(
+            inputRecord.getShapeErr()[0, 0],
+            outputRecord.getShapeErr()[0, 0], rtol=1e-6)
+        self.assertFloatsAlmostEqual(
+            inputRecord.getShapeErr()[1, 1],
+            outputRecord.getShapeErr()[1, 1], rtol=1e-6)
+        self.assertFloatsAlmostEqual(
+            inputRecord.getShapeErr()[2, 2],
+            outputRecord.getShapeErr()[2, 2], rtol=1e-6)
+
+    def testFitsCatalog(self):
+        storageClass = self.storageClassFactory.getStorageClass("SourceCatalog")
+        datasetTypeName = "testCatalog"
+
+        catalog = self.makeExampleCatalog()
+        dimensions = self.butler.registry.dimensions.extract(["instrument", "visit"])
+        addDatasetType(self.butler, datasetTypeName, dimensions, storageClass)
+
+        dataId = {"visit": 42, "instrument": "DummyCam", "physical_filter": "d-r"}
+        ref = self.butler.put(catalog, datasetTypeName, dataId)
+        stored = self.butler.get(ref)
+        self.assertCatalogEqual(catalog, stored)
+
+    def testExposureCompositePutGetConcrete(self):
+        storageClass = self.storageClassFactory.getStorageClass("ExposureF")
+        ref = self.runExposureCompositePutGetTest(storageClass, "calexp")
+
+        uri = self.butler.getURI(ref)
+        self.assertTrue(os.path.exists(uri.path), f"Checking URI {uri} existence")
+
+    def testExposureCompositePutGetVirtual(self):
+        storageClass = self.storageClassFactory.getStorageClass("ExposureCompositeF")
+        ref = self.runExposureCompositePutGetTest(storageClass, "unknown")
+
+        primary, components = self.butler.getURIs(ref)
+        self.assertIsNone(primary)
+        self.assertEqual(set(components), COMPONENTS)
+
+    def runExposureCompositePutGetTest(self, storageClass, datasetTypeName):
+        example = os.path.join(TESTDIR, "data", "small.fits")
+        exposure = lsst.afw.image.ExposureF(example)
+        dimensions = self.butler.registry.dimensions.extract(["instrument", "visit"])
+        addDatasetType(self.butler, datasetTypeName, dimensions, storageClass)
+
+        assembler = ExposureAssembler(storageClass)
```

Comment:
I suggest moving this to just above the loop body; it will be easier to understand what it's for (since the first `put`/`get` cycle doesn't use it at all).